lib/puppet.rb

```diff
@@ -205,8 +205,9 @@ def self.initialize_facts
 
     Facter.add(:agent_specified_environment) do
       setcode do
-        if Puppet.settings.set_by_config?(:environment)
-          Puppet[:environment]
+        if Puppet.settings.set_by_config?(:environment, nil, :agent)
+          # Look up the setting in the agent section first, then fallback to main
+          Puppet.settings.set_in_section?(:environment, :agent) || Puppet[:environment]
```

Comment:
`#set_in_section?` returns the value of a setting in the specified run mode or nil if it cannot be found

Reply:
The comment 'fallback to main' is a little misleading, because `Puppet[:environment]` includes `:memory` and sections relating to run mode. For example, if an environment is set in `main`, but not `agent`, and is overridden in memory, then the fallback to `Puppet[:environment]` will return the in-memory version:

```
❯ cat ~/.puppetlabs/etc/puppet/puppet.conf
[main]
environment=dev
❯ bx irb
irb(main):001:0> require 'puppet'
Pupp=> true
irb(main):002:0> Puppet.initialize_settings
=> {}
irb(main):003:0> Puppet::Util::Log.newdestination(:console)
=> #<Puppet::Util::Log::DestConsole:0x000055959a232118>
irb(main):004:0> Puppet[:environment]
=> "dev"
irb(main):005:0> Puppet[:environment] = 'from memory'
=> "from memory"
irb(main):006:0> Puppet.settings.set_by_config?(:environment, nil, :agent)
=> true
irb(main):007:0> Puppet.settings.set_in_section?(:environment, :agent)
=> nil
irb(main):008:0> Puppet.settings.set_in_section?(:environment, :main)
=> "dev"
irb(main):009:0> Puppet[:environment]
=> "from memory"
```

Similarly, if the environment is set to the `user` section and puppet is running in that run mode (as is the default):

```
❯ cat -p  ~/.puppetlabs/etc/puppet/puppet.conf
[main]
environment=dev

[user]
environment=in_user

❯ bx irb
irb(main):001:0> require 'puppet'
=> true
irb(main):002:0> Puppet.initialize_settings
=> {}
irb(main):003:0> Puppet::Util::Log.newdestination(:console)
=> #<Puppet::Util::Log::DestConsole:0x000055e5293804c8>
irb(main):004:0> Puppet.settings.set_by_config?(:environment, nil, :agent)
=> true
irb(main):005:0> Puppet.settings.set_in_section?(:environment, :agent)
=> nil
irb(main):006:0> Puppet.settings.set_in_section?(:environment, :main)
=> "dev"
irb(main):007:0> Puppet[:environment]
=> "in_user"
```

We could be more strict about the fallback behavior and only look at the `:main` section?

```ruby
Puppet.settings.set_in_section?(:environment, :agent) || \
Puppet.settings.set_in_section?(:environment, :main)
```

That said the fallback behavior was the same before this PR, so I'm fine not changing it.

Reply:
Hm yeah, good point @joshcooper. I'll update the code to only check `agent` and `main`.